### ISSUE TITLE
json-c/json_object.c:50:2: error: #error Unable to determine size of ssize_t

### ISSUE THREAD

**Issue by MarcT512** [json-c/json-c#638]
With latest master, make fails with "json-c/json_object.c:50:2: error: #error Unable to determine size of ssize_t"
The last commit that builds for me is around commit #4a546e7 - and the problem occurred sometime after that.

FYI - I can see that config.h has:
`#define SIZEOF_SSIZE_T 8`

...so I'm not sure why this is happening.

**Steps To Reproduce**
Build master on 64bit Linux 

**Version and Platform**
- json-c version: 34334e5d3f89fe7c1e07e776761f5fa614e67fb7
- OS: Slackware Linux 14.2 64bit & Slackware-current 64bit
- Custom cmake/build flags, if any: None

Full log:
```
marc@deepthought:~/git/json-c/build$ make
Scanning dependencies of target json-c
[  1%] Building C object CMakeFiles/json-c.dir/arraylist.c.o
[  2%] Building C object CMakeFiles/json-c.dir/debug.c.o
[  3%] Building C object CMakeFiles/json-c.dir/json_c_version.c.o
[  5%] Building C object CMakeFiles/json-c.dir/json_object.c.o
/home/marc/git/json-c/json_object.c:50:2: error: #error Unable to determine size of ssize_t
   50 | #error Unable to determine size of ssize_t
      |  ^~~~~
/home/marc/git/json-c/json_object.c: In function â_json_object_new_stringâ:
/home/marc/git/json-c/json_object.c:1269:13: error: âSSIZE_T_MAXâ undeclared (first use in this function); did you mean âSSIZE_MAXâ?
 1269 |  if (len > (SSIZE_T_MAX - (sizeof(*jso) - sizeof(jso->c_string)) - 1))
      |             ^~~~~~~~~~~
      |             SSIZE_MAX
/home/marc/git/json-c/json_object.c:1269:13: note: each undeclared identifier is reported only once for each function it appears in
/home/marc/git/json-c/json_object.c: In function â_json_object_set_string_lenâ:
/home/marc/git/json-c/json_object.c:1332:13: error: âSSIZE_T_MAXâ undeclared (first use in this function); did you mean âSSIZE_MAXâ?
 1332 |  if (len >= SSIZE_T_MAX - 1)
      |             ^~~~~~~~~~~
      |             SSIZE_MAX
make[2]: *** [CMakeFiles/json-c.dir/build.make:122: CMakeFiles/json-c.dir/json_object.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:236: CMakeFiles/json-c.dir/all] Error 2
make: *** [Makefile:183: all] Error 2
```

**Comment by MarcT512** [comment]
If it helps, I did a "git bisect" and the result is:

first bad commit: [0fc9d91277c24e7a9b8e2b0eeb4df16c2132ac2d] Kick json_type_string out of struct json_object. The default is now that string data is stored inline at the end of json_object, though to allow for json_object_set_string() to set a _longer_ string, we still need to allow for the possibility of a separate char * pointer. All json types have been split out now, next step it cleanup.

**Comment by MarcT512** [comment]
This diff fixes it for me on 64bit Linux. Not sure of the implications for Windows.

NB: There is an issue in the existing code in json_object.c. UINT_MAX, ULONG_MAX, ULLONG_MAX shouldn't be used to define the max size of a ssize_t, since these are unsigned maxima and ssize_t is a signed type. This is also fixed below.

```
diff --git a/json_object.c b/json_object.c
index 1dffd00..69ae283 100644
--- a/json_object.c
+++ b/json_object.c
@@ -39,13 +39,13 @@
 #error "The long long type isn't 64-bits"
 #endif
 
-#ifndef SSIZE_T_MAX
+#ifndef SSIZE_MAX
 #if SIZEOF_SSIZE_T == SIZEOF_INT
-#define SSIZE_T_MAX UINT_MAX
+#define SSIZE_MAX INT_MAX
 #elif SIZEOF_SSIZE_T == SIZEOF_LONG
-#define SSIZE_T_MAX ULONG_MAX
+#define SSIZE_MAX LONG_MAX
 #elif SIZEOF_SSIZE_T == SIZEOF_LONG_LONG
-#define SSIZE_T_MAX ULLONG_MAX
+#define SSIZE_MAX LLONG_MAX
 #else
 #error Unable to determine size of ssize_t
 #endif
@@ -1266,7 +1266,7 @@ static struct json_object *_json_object_new_string(const char *s, const size_t l
         *                       data
      *                       \0]
         */
-       if (len > (SSIZE_T_MAX - (sizeof(*jso) - sizeof(jso->c_string)) - 1))
+       if (len > (SSIZE_MAX - (sizeof(*jso) - sizeof(jso->c_string)) - 1))
                return NULL;
        objsize = (sizeof(*jso) - sizeof(jso->c_string)) + len + 1;
        if (len < sizeof(void *))
@@ -1329,7 +1329,7 @@ static int _json_object_set_string_len(json_object *jso, const char *s, size_t l
        if (jso == NULL || jso->o_type != json_type_string)
                return 0;
 
-       if (len >= SSIZE_T_MAX - 1)
+       if (len >= SSIZE_MAX - 1)
                // jso->len is a signed ssize_t, so it can't hold the
                // full size_t range.
                return 0;
```

**Comment by hawicz** [comment]
Why change the name of the define to not match the actual type?

**Comment by hawicz** [comment]
I think we need a better analysis of why your build was ending up at the #error line in the first place.  Did some of the SIZEOF_* values not get properly populated in config.h?  Are none of the integer types the same size as ssize_t on your platform?

Note that the change from the unsigned to signed constants *does* make sense to do; I'll apply that separately.

**Comment by MarcT512** [comment]
> Why change the name of the define to not match the actual type?

I changed the name because SSIZE_MAX is already defined on my system and it made sense to use it.
See https://www.gnu.org/software/libc/manual/html_node/General-Limits.html

> I think we need a better analysis of why your build was ending up at the #error line in the first place. Did some of the SIZEOF_* values not get properly populated in config.h? Are none of the integer types the same size as ssize_t on your platform?

Agreed - unfortunately I know very little about cmake. This is a standard 64bit x86_64 platform, so integer sizes should be completely standard. Slackware is a less common Linux distribution, but almost everything compiles fine from source with little or no changes.

config.h from the Slackware 14.2 64bit system looks like:

```
/* Enable RDRAND Hardware RNG Hash Seed */
/* #undef ENABLE_RDRAND */

/* Enable partial threading support */
/* #undef ENABLE_THREADING */

/* Define if .gnu.warning accepts long strings. */
/* #undef HAS_GNU_WARNING_LONG */

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H

/* Define to 1 if you have the <endian.h> header file. */
#define HAVE_ENDIAN_H

/* Define to 1 if you have the <fcntl.h> header file. */
#define HAVE_FCNTL_H

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H

/* Define to 1 if you have the <limits.h> header file. */
#define HAVE_LIMITS_H

/* Define to 1 if you have the <locale.h> header file. */
#define HAVE_LOCALE_H

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H

/* Define to 1 if you have the <stdarg.h> header file. */
#define HAVE_STDARG_H

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H

/* Define to 1 if you have the <syslog.h> header file. */
#define HAVE_SYSLOG_H 1

/* Define to 1 if you have the <sys/cdefs.h> header file. */
#define HAVE_SYS_CDEFS_H

/* Define to 1 if you have the <sys/param.h> header file. */
#define HAVE_SYS_PARAM_H 1

/* Define to 1 if you have the <sys/resource.h> header file. */
#define HAVE_SYS_RESOURCE_H

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the <xlocale.h> header file. */
#define HAVE_XLOCALE_H

/* Define to 1 if you don't have `vprintf' but do have `_doprnt.' */
/* #undef HAVE_DOPRNT */

/* Has atomic builtins */
#define HAVE_ATOMIC_BUILTINS

/* Define to 1 if you have the declaration of `INFINITY', and to 0 if you
   don't. */
#define HAVE_DECL_INFINITY

/* Define to 1 if you have the declaration of `isinf', and to 0 if you don't.
   */
#define HAVE_DECL_ISINF

/* Define to 1 if you have the declaration of `isnan', and to 0 if you don't.
   */
#define HAVE_DECL_ISNAN

/* Define to 1 if you have the declaration of `nan', and to 0 if you don't. */
#define HAVE_DECL_NAN

/* Define to 1 if you have the declaration of `_finite', and to 0 if you
   don't. */
/* #undef HAVE_DECL__FINITE */

/* Define to 1 if you have the declaration of `_isnan', and to 0 if you don't.
   */
/* #undef HAVE_DECL__ISNAN */

/* Define to 1 if you have the `open' function. */
#define HAVE_OPEN

/* Define to 1 if you have the `realloc' function. */
#define HAVE_REALLOC

/* Define to 1 if you have the `setlocale' function. */
#define HAVE_SETLOCALE

/* Define to 1 if you have the `snprintf' function. */
#define HAVE_SNPRINTF


/* Define to 1 if you have the `strcasecmp' function. */
#define HAVE_STRCASECMP 1

/* Define to 1 if you have the `strdup' function. */
#define HAVE_STRDUP

/* Define to 1 if you have the `strerror' function. */
#define HAVE_STRERROR

/* Define to 1 if you have the `strncasecmp' function. */
#define HAVE_STRNCASECMP 1

/* Define to 1 if you have the `uselocale' function. */
#define HAVE_USELOCALE

/* Define to 1 if you have the `vasprintf' function. */
#define HAVE_VASPRINTF

/* Define to 1 if you have the `vprintf' function. */
#define HAVE_VPRINTF

/* Define to 1 if you have the `vsnprintf' function. */
#define HAVE_VSNPRINTF

/* Define to 1 if you have the `vsyslog' function. */
#define HAVE_VSYSLOG 1

/* Define if you have the `getrusage' function. */
#define HAVE_GETRUSAGE

#define HAVE_STRTOLL
#if !defined(HAVE_STRTOLL)
#define strtoll strtoll
/* #define json_c_strtoll strtoll*/
#endif

#define HAVE_STRTOULL
#if !defined(HAVE_STRTOULL)
#define strtoull strtoull
/* #define json_c_strtoull strtoull */
#endif

/* Have __thread */
#define HAVE___THREAD

/* Public define for json_inttypes.h */
#define JSON_C_HAVE_INTTYPES_H 1

/* Name of package */
#define PACKAGE "json-c"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "json-c@googlegroups.com"

/* Define to the full name of this package. */
#define PACKAGE_NAME "json-c"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "json-c 0.14.99"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "json-c"

/* Define to the home page for this package. */
#define PACKAGE_URL "https://github.com/json-c/json-c"

/* Define to the version of this package. */
#define PACKAGE_VERSION "0.14.99"

/* The number of bytes in type int */
#define SIZEOF_INT 4

/* The number of bytes in type int64_t */
#define SIZEOF_INT64_T 8

/* The number of bytes in type long */
#define SIZEOF_LONG 8

/* The number of bytes in type long long */
#define SIZEOF_LONG_LONG 8

/* The number of bytes in type size_t */
#define SIZEOF_SIZE_T 8

/* The number of bytes in type ssize_t */
#define SIZEOF_SSIZE_T 8

/* Specifier for __thread */
#define SPEC___THREAD __thread

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS

/* Version number of package */
#define VERSION "0.14.99"

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */

/* Define to `unsigned int' if <sys/types.h> does not define. */
/* #undef size_t */
```

**Comment by hawicz** [comment]
Forget about cmake, it's irrelevant.  By the time you're compiling json_object.c all that matters is what's in the .c and .h files.  The relevant code in config.h starts at:
```
/* The number of bytes in type int */
#define SIZEOF_INT 4
...
```
and it seems like all the right values are there so the comparisons in json_object.c that do `#if SIZEOF_SSIZE_T == SIZEOF_INT`` etc... should work just fine, and on your system should match on the `#elif SIZEOF_SSIZE_T == SIZEOF_LONG` line.  I suggest using a technique like described at https://stackoverflow.com/questions/1562074/how-do-i-show-the-value-of-a-define-at-compile-time and add the following to your json_object.c:

```
#define STR(x) #x
#define XSTR(x) STR(x)
#pragma message "int:"  XSTR(SIZEOF_INT)
#pragma message "long:"  XSTR(SIZEOF_LONG)
#pragma message "ssize_t:"  XSTR(SIZEOF_SSIZE_T)
```

**Comment by MarcT512** [comment]
Thanks for the hint. Here's the output - looks like SIZEOF_SSIZE_T is not being defined:

```
marc@deepthought:~/git/json-c/build$ make
Scanning dependencies of target json-c
[  1%] Building C object CMakeFiles/json-c.dir/json_object.c.o
/home/marc/git/json-c/json_object.c:44:9: note: ‘#pragma message: int:4’
   44 | #pragma message "int:"  XSTR(SIZEOF_INT)
      |         ^~~~~~~
/home/marc/git/json-c/json_object.c:45:9: note: ‘#pragma message: long:8’
   45 | #pragma message "long:"  XSTR(SIZEOF_LONG)
      |         ^~~~~~~
/home/marc/git/json-c/json_object.c:46:9: note: ‘#pragma message: ssize_t:SIZEOF_SSIZE_T’
   46 | #pragma message "ssize_t:"  XSTR(SIZEOF_SSIZE_T)
      |         ^~~~~~~
/home/marc/git/json-c/json_object.c:56:2: error: #error Unable to determine size of ssize_t
   56 | #error Unable to determine size of ssize_t
      |  ^~~~~
/home/marc/git/json-c/json_object.c: In function ‘_json_object_new_string’:
/home/marc/git/json-c/json_object.c:1275:13: error: ‘SSIZE_T_MAX’ undeclared (first use in this function); did you mean ‘SSIZE_MAX’?
 1275 |  if (len > (SSIZE_T_MAX - (sizeof(*jso) - sizeof(jso->c_string)) - 1))
      |             ^~~~~~~~~~~
      |             SSIZE_MAX
/home/marc/git/json-c/json_object.c:1275:13: note: each undeclared identifier is reported only once for each function it appears in
/home/marc/git/json-c/json_object.c: In function ‘_json_object_set_string_len’:
/home/marc/git/json-c/json_object.c:1338:13: error: ‘SSIZE_T_MAX’ undeclared (first use in this function); did you mean ‘SSIZE_MAX’?
 1338 |  if (len >= SSIZE_T_MAX - 1)
      |             ^~~~~~~~~~~
      |             SSIZE_MAX
make[2]: *** [CMakeFiles/json-c.dir/build.make:122: CMakeFiles/json-c.dir/json_object.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:236: CMakeFiles/json-c.dir/all] Error 2
make: *** [Makefile:183: all] Error 2
```

**Comment by MarcT512** [comment]
Ok - I figured it out, I think:

I have a config.h in ./json-c AND a different (correct) one created by cmake in ./json-c/build
```
marc@deepthought:~/git/json-c$ find . -iname 'config.h' -ls
  1201062      8 -rw-r--r--   1 marc     users        5342 Feb 13  2019 ./config.h
  1182334      8 -rw-r--r--   1 marc     users        5565 Jul  1 22:15 ./build/config.h
```

The one in ./json-c does not have SIZEOF_SSIZE_T defined, so the compiler must be including the wrong one.
The same problem does occur if I built inside the source tree or external to it, as I tested that before opening the issue.

I suspect the "old" one was left over after the switch from autoconf to cmake, and since it's also listed in .gitignore I didn't see it in "git status". I wonder how many other people have/will run into the same issue...?
